### PR TITLE
Carrier can take a shipment offer

### PR DESCRIPTION
**Pull request by matzayonc**
closeses #17

### CODE REVIEW

File: programs/protocol/src/data/shipper.rs [hunk 1]
```diff
@@ -1,8 +1,12 @@
 use anchor_lang::prelude::*;
 
+use crate::Name;
+
 #[account(zero_copy)]
 #[derive(Debug, Default, PartialEq, Eq)]
 pub struct Shipper {
+    pub creator: Pubkey,
```
Comment (tommysr) [inline programs/protocol/src/data/shipper.rs:8]: Who is creator?

File: programs/protocol/src/data/name.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+use anchor_lang::prelude::{
+    borsh::{BorshDeserialize, BorshSerialize},
+    *,
+};
+
+#[zero_copy]
+#[derive(Debug, Default, PartialEq, BorshSerialize, BorshDeserialize, Eq)]
+pub struct Name {
+    pub value: [u32; 16], // 64 characters
```
Comment (tommysr) [inline programs/protocol/src/data/name.rs:9]: isnt better to store u16 like utf8 in rust?

File: programs/protocol/src/data/carrier.rs [hunk 3]
```diff
@@ -15,6 +15,10 @@ pub struct Availability {
 #[account(zero_copy)]
 #[derive(Debug, Default, PartialEq, BorshSerialize, BorshDeserialize)]
 pub struct Carrier {
+    pub creator: Pubkey,
     pub authority: Pubkey,
+    pub name: Name,
     pub availability: Availability,
+    pub offers: u32,
```
Comment (tommysr) [inline programs/protocol/src/data/carrier.rs:22]: count lost meaning i guess

File: programs/protocol/src/data/shipper.rs (hunk 1 above)
Comment (matzayonc) [reply programs/protocol/src/data/shipper.rs:8]: the creator of the shipper account, the one that is a part of seeds

File: programs/protocol/src/data/name.rs (hunk 2 above)
Comment (matzayonc) [reply programs/protocol/src/data/name.rs:9]: not sure why but it would not compile

File: programs/protocol/src/data/carrier.rs (hunk 3 above)
Comment (matzayonc) [reply programs/protocol/src/data/carrier.rs:22]: offers are offers to accept shipments 
count is a count of accepted shipments

File: programs/protocol/src/data/offer.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+use anchor_lang::prelude::*;
+
+use crate::ShipmentData;
+
+#[zero_copy]
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct OfferDetails {
+    pub payment: u64,
+    pub collateral: u64,
```
Comment (tommysr) [inline programs/protocol/src/data/offer.rs:9]: what's the difference?

File: programs/protocol/src/data/shipper.rs (hunk 1 above)
Comment (tommysr) [reply programs/protocol/src/data/shipper.rs:8]: But isn't it the same as authority in all cases?

**Review by tommysr — APPROVED** [review]
gj

File: programs/protocol/src/data/carrier.rs (hunk 3 above)
Comment (tommysr) [reply programs/protocol/src/data/carrier.rs:22]: but isnt better to call it offers_count accepted_offers_count or smth

Comment (tommysr) [reply programs/protocol/src/data/carrier.rs:22]: count is meaningless

File: programs/protocol/src/data/shipper.rs (hunk 1 above)
Comment (matzayonc) [reply programs/protocol/src/data/shipper.rs:8]: so far yes, but doesn't have to be in the future, it allows us to add a delegate to sign transaction for another account
It's not necessary at this point in development, but decided to differentiate between signature verification and seeds

File: programs/protocol/src/data/carrier.rs (hunk 3 above)
Comment (matzayonc) [reply programs/protocol/src/data/carrier.rs:22]: right, will do

File: programs/protocol/src/data/offer.rs (hunk 4 above)
Comment (matzayonc) [reply programs/protocol/src/data/offer.rs:9]: payment is whatever the truct driver will get for the transport
collateral is how much he will have to stake as collateral for shipment value

**Pull request merged**